node/src/components/consensus/protocols/highway.rs

```diff
@@ -299,14 +299,22 @@ impl<I: NodeIdT, C: Context + 'static> HighwayProtocol<I, C> {
                             if let (Some(value), Some(timestamp)) =
                                 (vertex.value().cloned(), vertex.timestamp())
                             {
-                                // It's a block: Request validation before adding it to the state.
-                                self.pending_values
-                                    .entry(value.clone())
-                                    .or_default()
-                                    .push(vv);
-                                results.push(ProtocolOutcome::ValidateConsensusValue(
-                                    sender, value, timestamp,
-                                ));
+                                // It's a block…
+                                if value.is_empty() {
+                                    // Consensus value is empty. Nothing to validate.
+                                    let now = Timestamp::now();
+                                    results.extend(self.add_valid_vertex(vv, rng, now));
+                                    state_changed = true;
```

Comment:
Can this be deduplicated and merged with the `else` below?
```rust
match (vertex.value().cloned(), vertex.timestamp()) {
    (Some(value), Some(timestamp)) if !value.is_empty() => {
        // Request validation before adding it to the state.
        // …
    }
    (_, _) => {
        // No validation required. Add to the state.
        // …
    }
}
```
Or does the borrow checker complain about that?

Reply:
I haven't really tried. I did see the duplication but I thought I will go with the simplicity for the reader's sake.

Reply:
Done in [`09ae779` (#653)](https://github.com/CasperLabs/casper-node/pull/653/commits/09ae779c37005427481d73af152128e242604999)